### PR TITLE
Use Notification Service verification for emails

### PR DESCRIPTION
**Pull request by wilpow-github**
### Context

Our email address validation is fairly permissive and it's allowing some emails which are subsequently failing when we try to email them via Notify. When this happens, the Hangfire job fails (and retries then fails again...) but users aren't informed that their email is invalid.

### Changes proposed in this pull request

Make the Notify API call synchronously rather than from a background job.
Modify the email verification process to capture the 'invalid email' error from Notify. 

### Checklist

-   [ ] Attach to Trello card
-   [ ] Rebased master
-   [ ] Cleaned commit history
-   [ ] Tested by running locally
-   [ ] Reminder created to manually clean any removed app settings post deployment

### CODE REVIEW

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Email.cshtml.cs [hunk 1]
```diff
@@ -41,43 +33,15 @@ public async Task<IActionResult> OnPost()
...
-            var existingUser = await _dbContext.Users.Where(user => user.EmailAddress == Email).SingleOrDefaultAsync();
-            if (existingUser is null)
-            {
-                ModelState.AddModelError(nameof(Email), "Enter a personal email address. It cannot be one that other people may get access to.");
-                return this.PageWithErrors();
-            }
+            return validateEmailResult.Result!;
         }
 
         HttpContext.GetAuthenticationState().OnEmailSet(Email!);
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Email.cshtml.cs:43]: This now happens after the pin generation step. I think that's fine

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Email.cshtml.cs [hunk 2]
```diff
@@ -1,28 +1,20 @@
...
-        TeacherIdentityServerDbContext dbContext)
+        TeacherIdentityServerDbContext dbContext) :
+        base(emailVerificationService, linkGenerator, dbContext)
     {
-        _emailVerificationService = emailVerificationService;
-        _linkGenerator = linkGenerator;
-        _dbContext = dbContext;
     }
 
     [Display(Name = "Enter your email address", Description = "Use your personal email address. This is so you can keep these sign in details should you change jobs.")]
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Email.cshtml.cs:20]: The email input has slightly different descriptions and so can't be commonised

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+            case PinGenerationFailedReasons.RateLimitExceeded:
+                return EmailValidationResult.Failed(new ViewResult()
+                {
+                    StatusCode = 429,
+                    ViewName = "TooManyRequests"
+                });
+
+            case PinGenerationFailedReasons.InvalidEmail:
+                ModelState.AddModelError(nameof(email), "Enter a valid email address");
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs:57]: Generic validation failure text

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                ModelState.AddModelError(nameof(email), "Enter a personal email address. It cannot be one that other people may get access to.");
+                return EmailValidationResult.Failed(this.PageWithErrors());
+            }
+        }
+
+        return await TryGeneratePinForEmail(email);
+    }
+
+    public async Task<EmailValidationResult> TryGeneratePinForEmail(string email)
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs:40]:
```suggestion
    private async Task<EmailValidationResult> TryGeneratePinForEmail(string email)
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.EntityFrameworkCore;
+using TeacherIdentity.AuthServer.Models;
+using TeacherIdentity.AuthServer.Services.EmailVerification;
+
+namespace TeacherIdentity.AuthServer.Pages.SignIn;
+
+public class BaseEmailPageModel : PageModel
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs:9]:
```suggestion
public async class BaseEmailPageModel : PageModel
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        IEmailVerificationService emailVerificationService,
+        IIdentityLinkGenerator linkGenerator,
+        TeacherIdentityServerDbContext dbContext)
+    {
+        _emailVerificationService = emailVerificationService;
+        LinkGenerator = linkGenerator;
+        _dbContext = dbContext;
+    }
+
+    public async Task<EmailValidationResult> TryValidateEmail(string email)
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/BaseEmailPageModel.cs:25]:
```suggestion
    public async Task<EmailValidationResult> ValidateEmail(string email)
```

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Services/EmailVerification/EmailVerificationService.cs [hunk 7]
```diff
@@ -154,4 +162,12 @@ public async Task<PinVerificationFailedReasons> VerifyPin(string email, string p
 
         return PinVerificationFailedReasons.None;
     }
+
+    private const string EmailSubject = "Confirm your email address";
+    private string EmailBodyTemplate(string pin, int minutes, string clientName) => $"Use this code to confirm your email address:\n\n" +
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Services/EmailVerification/EmailVerificationService.cs:167]:
```suggestion
    private string GetEmailBody(string pin, int minutes, string clientName) => $"Use this code to confirm your email address:\n\n" +
```

**Review by gunndabad — APPROVED** [review]

**Pull request merged**